src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -159,6 +159,15 @@ public bool CanRetry
                 if (values.Length > 0)
                 {
                     await WriteStringAsync(values[0], cancellationToken).ConfigureAwait(false);
+
+                    if (cookies != null && header.Key == HttpKnownHeaderNames.Cookie)
+                    {
+                        await WriteTwoBytesAsync((byte)';', (byte)' ', cancellationToken).ConfigureAwait(false);
+                        await WriteStringAsync(cookies, cancellationToken).ConfigureAwait(false);
+
+                        cookies = null;
+                    }
```

Comment:
Will a Cookie header never have multiple values?  I'm wondering about the loop below this... seems like if it did have multiple values, they'll end up getting appended with commas.

Reply:
A 'Cookie' request header can have multiple cookies on the single 'Cookie' request header.

There has been confusion on this over the years.  The latest RFC 6265 states:
https://tools.ietf.org/html/rfc6265#section-5.4

"5.4 When the user agent generates an HTTP request, the user agent MUST NOT attach more than one Cookie header field."

Reply:
The various HTTP APIs we have allow for multiple Cookie request headers to be set.  But the end result of the HTTP stack is supposed to only send a single 'Cookie' header on the wire for the request.

Reply:
Yeah, basically the Cookie header is weird because Netscape designed it poorly.

If the user sets multiple Cookie headers, that's arguably an error on their part.  We could handle this by appending with ";" instead of "," -- this should be easy to do since we're already doing special handling here.  Or we could just leave as-is and tell them not to do this.

I'm not sure what the other handlers do here; I'll investigate.

Note there's no strongly-typed Cookie property on the HttpRequestHeaders.  Presumably this is intentional...

Reply:
To be clear, though, I wasn't meaning to suggest multiple cookie headers, but rather multiple values for the single header, e.g.
```C#
using System;
using System.Net.Http;
using System.Threading.Tasks;

class Program
{
    static async Task Main()
    {
        var m = new HttpRequestMessage();
        m.Headers.Add("Cookie", new[] { "CustomCookie=456", "AnotherCustomCookie=789" });
        Console.WriteLine(m.Headers.ToString());
        foreach (var c in m.Headers.GetValues("Cookie"))
        {
            Console.WriteLine(c);
        }
    }
}
```
outputs:
```
Cookie: CustomCookie=456, AnotherCustomCookie=789

CustomCookie=456
AnotherCustomCookie=789
```
Am I misunderstanding?

Reply:
Yes, I meant multiple values.

If the user sets multiple values for a Cookie header, that's arguably an error on their part. We could handle this by appending with ";" instead of "," -- this should be easy to do since we're already doing special handling here. Or we could just leave as-is and tell them not to do this.

I'll investigate what other handlers do here.

Reply:
Ok, thanks.

Reply:
>Note there's no strongly-typed Cookie property on the HttpRequestHeaders. Presumably this is intentional...

Yes.  It is intentional.  The HttpWebRequest and HttpClient/HttpClientHandler APIs encouraged the use of the CookieContainer.  Overall having devs manually add 'Cookie' request headers is discouraged.

Reply:
WinHttpHandler behaves essentially the same as current SocketsHttpHandler here.  That is, if you specify multiple Cookie headers, they get appended using a comma, e.g. "A=1, B=2".  If you also have a cookie from the container, you get something like "ContainerCookie=Foo; A=1, B=2" -- that is, the container cookie is added to one of the cookie headers.

I'm fine with this behavior, as it matches header handling generally.  I suppose we could special-case manual cookie header handling, but I don't think there's any real value here.  Thoughts?

Reply:
Sounds fine then.